hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -409,6 +366,64 @@ def draw_bytes(
     ) -> bytes:
         raise NotImplementedError
 
+    def per_test_case_context_manager(self):
+        return contextlib.nullcontext()
+
+    def realize(self, value: T, *, for_failure: bool = False) -> T:
+        """
+        Called whenever hypothesis requires a concrete (non-symbolic) value from
+        a potentially symbolic value. Hypothesis will not check that `value` is
+        symbolic before calling `realize`, so you should handle the case where
+        `value` is non-symbolic.
+
+        The returned value should be non-symbolic.  If you cannot provide a value,
+        raise hypothesis.errors.BackendCannotProceed("discard_test_case").
+
+        If for_failure is True, the value is associated with a failing example.
+        In this case, the backend should spend substantially more effort when
+        attempting to realize the value, since it is important to avoid discarding
+        failing  examples. Backends may still raise BackendCannotProceed when
+        for_failure is True, if realization is truly impossible or if realization
+        takes significantly longer than expected (say, 5 minutes).
+        """
+        return value
+
+    def replay_choices(self, choices: tuple[ChoiceT, ...]) -> None:
+        """
+        Called to indicate that the provider should replay this choice sequence
+        the next time Hypothesis enters per_test_case_context_manager, instead of
+        generating its own choice sequence.
+
+        This is intended to indicate that the passed choice sequence is particularly
+        interesting in some way, and that the provider should collect runtime
+        information about it while replaying, in order to generate more effective
+        choice sequences in the future.
+
+        This method is currently used for relatively tight integration between
+        Hypothesis and specific providers.
```

Comment:
```suggestion
        Called when Hypothesis has discovered a choice sequence which the provider
        may wish to enqueue to replay under its own instrumentation when we next
        ask to generate a test case, rather than generating one from scratch.

        This is used to e.g. warm-start hypothesis-crosshair with a corpus of
        high-code-coverage inputs discovered by hypofuzz.
```
I think we can be concrete and specific here.